docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,234 @@ let f = (arg1: num, var arg2: num) => {
+* The `if let y = opt { } else { }` is a special if statement that binds `y` inside the block only if `x` includes a value. Otherwise, the `else` block will be executed.
 
-<details><summary>Equivalent TypeScript Code</summary>
+-------------
+* The `let x = opt else { }` can be used to <<<<<<>>>>>>???
+-------------
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
-> ```
+* The `x?.y` notation can be used to access object fields only if there is a value. It returns an empty optional if the value is not defined.
+* The `x ?? y` notation will return the value in `x` (if there is one) or `y` otherwise (`y` must be of type `T`, not `T?`).
+* A default value can also be supplied using `= y` in the declaration of the field or argument, which implies that the type will always be defined.
 
-</details>
+#### 1.7.1 Declaration
+
+##### 1.7.1.1 Declaration in struct fields
+
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+assert(my_person.address? == false);
+```
+
+It is also possible to use `=` in the field declaration to indicate the default value (similar to function arguments). If a default value is provided in a struct declaration, the field is not required in a struct literal definition. It also implies that the type of the field is `T` and not `T?` (because we can ensure it has a value). In the example below the field `radix` as a type of `num`.
+
+```js
+struct FormatOpts {
+  radix: num = 10;
+  some_optional: str?;
+}
+
+let opts = FormatOpts {};
+assert(opts.radix == 10);
+assert(opts.some_optional? == false); // <-- no value inside `some_optional`
+```
+
+A value can be omitted from a struct literal if the field is optional _or_ if it has a default value in the struct
+declaration. If the field is optional, it's type is `T?`. If it has a default value it's type is `T` (since we know it will always have a value).
+In the above example the type of `radix` is `num`.
+
+##### 1.7.1.2 Declaration in variables
+
+Use `T?` to indicate that a variable is optional:
+
+```js
+let var y: num?;
+assert(y? == false); // y doesn't have a value
+
+// ok to reassign another value because `y` is reassignable (`var`)
+y = 123;
+assert(y? == true);
+```
+
+It is okay to initialize a value:
+
+```js
+let var y: num? = 12;
+```
+
+##### 1.7.1.3 Declaration in class fields
+
+Similarly to struct fields, fields of classes can be also defined as optional using `T?`:
+
+```js
+class Foo {
+  my_opt: num?;
+  var my_var: str?;
+
+  init(opt: num?) {
+    this.my_opt = opt;
+  }
+
+  set_my_var(x: str) {
+    this.my_var = x;
+  }
+}
+```
```

Comment:
I was thinking about resetting:
```js
class Foo {
  var my_var1: num?;
  var my_var2: str?;

  set_my_var1(x: num?) {
    this.my_var1 = x;
  }

  set_my_var2(x: str) {
    this.my_var2 = x;
  }
  
  reset() {
    // my_var1 and my_var2 are supposed to be null again, but how?
  }
}
```
I guess I can do some dumb thing:
```js
class Foo {
  var my_var1: num?;
  var my_var2: str?;

  set_my_var1(x: num?) {
    this.my_var1 = x;
  }

  set_my_var2(x: str?) {
    this.my_var2 = x;
  }
  
  reset() {
    let new_foo = new Foo();
    this.my_var1 = new_foo.my_var1;
    this.my_var2 = new_foo.my_var2;
  }
}
```
But this definitely makes me feel I'm walking around some language limitations

Reply:
Reading through the paragraphs below, I found I can do something a bit more nature:
```js
class Foo {
  var my_var1: num?;
  var my_var2: str?;

  set_my_var1(x: num?) {
    this.my_var1 = x;
  }

  set_my_var2(x: str?) {
    this.my_var2 = x;
  }

  reset_my_var1(): num? { // still weird to have return type, then I will need more code to do the right encapsulation
    return nil;
  }

  reset_my_var2(): str? {
    return nil;
  }
  
  reset() {
    this.my_var1 = this.reset_my_var1();
    this.my_var2 = this.reset_my_var2();
  }
}
```